lnst/Recipes/ENRT/ConfigMixins/LongLivedConnectionsMixin.py

```diff
@@ -0,0 +1,230 @@
+
+    def generate_perf_endpoints(self, config):
+        return self._filter_ip_endpoints(
+            config, slice(0, 1)  # only the first IP is used for perf
+        )
+
+    def generate_long_lived_conns_endpoints(self, config):
+        return self._filter_ip_endpoints(
+            config,
+            slice(1, None),  # all IPs except the first one which is used for perf
+        )
+
+    @property
+    def servers_count(self):
+        return ceil(self.params.long_lived_conns / self.params.long_lived_conns_per_ip)
+
+    def calculate_client_connections(self, client_id):
+        client_id += 1  # 0-based index
+        if client_id < self.servers_count:
+            return self.params.long_lived_conns_per_ip
+
+        # remaining connections are handled by last client
+        return (
+            self.params.long_lived_conns
+            - (self.servers_count - 1) * self.params.long_lived_conns_per_ip
+        )
+
+    def _prepare_server(self, receiver, conns_count):
+        server = LongLivedServer(
+            server_ip=receiver.address,
+            server_port=self.params.long_lived_conns_port,
+            connections_count=conns_count,
+        )
+
+        job = receiver.device.netns.prepare_job(server)
+
+        return job
+
+    def _prepare_client(self, generator, receiver, conns_count):
+        client = LongLivedClient(
+            server_ip=receiver.address,
+            server_port=self.params.long_lived_conns_port,
+            client_ip=generator.address,
+            connections_count=conns_count,
+        )
+
+        job = generator.device.netns.prepare_job(client)
+
+        return job
+
+    def wait_agent_for_establish_conns(self, agent, timeout=60):
+        # Yeah, a bit hacky but simplest way of waiting.
+        # The problem is, that LNST doesn't support waiting
+        # for a specific condition at a REMOTE agent.
+
+        # TODO: refactor to use condition wait, when implemented
+        while timeout:
+            job = agent.run("ss | grep ESTAB | wc -l")
+            conns = int(job.stdout)
+            logging.debug(
+                f"Connections established: {conns}/{self.params.long_lived_conns}"
+            )
+            if conns > self.params.long_lived_conns:
+                logging.info("All long-lived connections established")
+                break
+            time.sleep(1)
+            timeout -= 1
+
+    def test_wide_configuration(self):
+        config = super().test_wide_configuration()
+        # L4 can address up to 65535 ports (size of addresable space
+        # defined in long_lived_conns_per_ip) Therefore opening connections
+        # may require multiple IPs.
+        host1, host2 = self.matched.host1, self.matched.host2
+
+        ipv4_addr = interface_addresses(self.params.net_ipv4)
+        ipv6_addr = interface_addresses(self.params.net_ipv6)
+
+        for _ in range(2):
+            # 2 addresses are already assigned by super().test_wide_configuration()
```

Comment:
**any** number of addresses may be assigned by the base Recipe, this uses an assumption that that equals 2, but technically we should probably reuse the ip address network iterator from the base Recipe - e.g. it can be passed via the `config` object:
```
BaseRecipe():
    def testwide():
        config.ipv4_addr_net = interface_addresses(self.params.net_ipv4)

LongLivedConnectionsMixin:
    def testwide():
        config = super().testwide()
        ipv4_addr = config.ipv4_addr_net
```

I realize this requires updating the base recipes that this mixin can be applied to, maybe there's a way to detect the "last used" address in a different way, but this is the "clean" way to implement this.

Reply:
Oh, thats something i didn't count with. Thanks.

I've added a separate `long_lived_conns_net{4,6}` params.